src/Voron/StorageEnvironment.cs

```diff
@@ -38,8 +38,49 @@ namespace Voron
 {
     public delegate bool UpgraderDelegate(SchemaUpgradeTransactions transactions, int currentVersion, out int versionAfterUpgrade);
 
-    public class StorageEnvironment : IDisposable
+
+    public class StorageEnvironmentSynchronization
     {
+        public struct SynchronizationSection : IDisposable
+        {
+            private readonly Semaphore _context;
+
+            public SynchronizationSection(Semaphore context)
+            {
+                this._context = context;
+                if (this._context != null)
+                    this._context.WaitOne();
+            }
+
+            void IDisposable.Dispose()
+            {
+                if (this._context != null)
+                    _context.Release();
+            }
+        }
+
+        private readonly Semaphore _writeSemaphore;
+        public readonly long MaxWriteSize;
+
+        public StorageEnvironmentSynchronization()
+        {
+            this.MaxWriteSize = int.MaxValue;
+        }
+
+        public StorageEnvironmentSynchronization(int concurrentWrites, long maxWriteSize = long.MaxValue)
+        {
+            this._writeSemaphore = new Semaphore(concurrentWrites, concurrentWrites);
+            this.MaxWriteSize = maxWriteSize;
+        }        
+
+        public SynchronizationSection Enter()
```

Comment:
We are going to need to plug this into the I/O metrics, so we can see how long we waited for this.